extensions/amp-auto-lightbox/0.1/carousel-criteria.js

```diff
@@ -0,0 +1,294 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {Layout} from '../../../src/layout';
+import {Services} from '../../../src/services';
+import {dev, devAssert} from '../../../src/log';
+import {isAmpElement, iterateCursor} from '../../../src/dom';
+import {isExperimentOn} from '../../../src/experiments';
+import {resolveFalse, resolveTrue} from './utils/promise';
+import {toArray, toWin} from '../../../src/types';
+import {tryResolve} from '../../../src/utils/promise';
+
+const MAX_TRAVERSING_DEPTH = 8;
+const MAX_TRAVERSING_BREADTH = 8;
+
+const MIN_IMG_SLIDE_AREA_RATIO = 0.5;
+
+/** @private @const {!Object<string, boolean>} */
+const ADMISSIBLE_AMP_ELEMENTS = {
+  'AMP-FIT-TEXT': true,
+  'AMP-IMG': true,
+  'AMP-LAYOUT': true,
+};
+
+/** @private @const {!Object<string, boolean>} */
+const ADMISSIBLE_HTML_ELEMENTS = {
+  'A': true,
+  'ABBR': true,
+  'ACRONYM': true,
+  'B': true,
+  'BDI': true,
+  'BDO': true,
+  'BIG': true,
+  'BLOCKQUOTE': true,
+  'BR': true,
+  'CAPTION': true,
+  'CENTER': true,
+  'CITE': true,
+  'DATA': true,
+  'DATALIST': true,
+  'DIV': true,
+  'EM': true,
+  'FIGCAPTION': true,
+  'FIGURE': true,
+  'H1': true,
+  'H2': true,
+  'H3': true,
+  'H4': true,
+  'H5': true,
+  'H6': true,
+  'HGROUP': true,
+  'HR': true,
+  'I': true,
+  'IMG': true,
+  'LABEL': true,
+  'LEGEND': true,
+  'NOBR': true,
+  'NOSCRIPT': true,
+  'O:P': true,
+  'P': true,
+  'PRE': true,
+  'Q': true,
+  'RP': true,
+  'RT': true,
+  'RTC': true,
+  'RUBY': true,
+  'S': true,
+  'SCRIPT': true,
+  'SECTION': true,
+  'SMALL': true,
+  'SPAN': true,
+  'STRIKE': true,
+  'STRONG': true,
+  'SUB': true,
+  'SUMMARY': true,
+  'SUP': true,
+  'TIME': true,
+  'U': true,
+  'VAR': true,
+  'WBR': true,
+
+  // The following could potentially be admissible but are either unlikely
+  // to be used where auto-lightbox is desired, or are obsolete.
+  // This list is kept but commented-out to keep track of possible elements to
+  // consider.
+  // 'ADDRESS': true,
+  // 'ARTICLE': true,
+  // 'ASIDE': true,
+  // 'CODE': true,
+  // 'DD': true,
+  // 'DEL': true,
+  // 'DFN': true,
+  // 'DL': true,
+  // 'DT': true,
+  // 'FOOTER': true,
+  // 'HEADER': true,
+  // 'INS': true,
+  // 'KBD': true,
+  // 'LISTING': true,
+  // 'MAIN': true,
+  // 'MARK': true,
+  // 'METER': true,
+  // 'MULTICOL': true,
+  // 'NAV': true,
+  // 'NEXTID': true,
+  // 'SAMP': true,
+  // 'SPACER': true,
+};
+
+export class CarouselCriteria {
+  /**
+   * @param {!Element} element
+   * @return {!Promise<boolean>}
+   */
+  static meetsAll(element) {
+    const win = toWin(element.ownerDocument.defaultView);
+
+    if (!isExperimentOn(win, 'amp-auto-lightbox-carousel')) {
+      return resolveFalse();
+    }
+
+    const slides = element.querySelectorAll('.amp-carousel-slide');
```

Comment:
there is a race condition since it carousel extension is not loaded yet, we don't have this class.

Reply:
Nope, this runs after `layoutCallback`.